Upgrade to v3.14.9 and remove okhttp3 url-connection (#3)
jenkinsci/okhttp-api-plugin: README.adoc

# OkHttp API Plugin for Jenkins

This plugin manages okhttp library version for Jenkins plugins.

> It includes the following packages:

Reviewer: Maybe worth noting explicitly that this plugin does not and will not include `okhttp-urlconnection` for OkHttp 3.x going forward.

azure-artifact-manager: okhttp-3.14.1.jar
azure-batch-parallel: okhttp-3.3.1.jar
azure-batch-parallel: okhttp-urlconnection-3.3.1.jar
azure-commons: okhttp-3.12.0.jar
azure-commons: okhttp-urlconnection-3.11.0.jar
azure-container-agents: okhttp-3.11.0.jar
azure-container-agents: okhttp-urlconnection-3.11.0.jar
azure-container-registry-tasks: okhttp-3.12.0.jar
azure-container-registry-tasks: okhttp-urlconnection-3.11.0.jar
azure-credentials-ext: okhttp-3.12.6.jar
azure-credentials: okhttp-3.12.6.jar
azure-credentials: okhttp-urlconnection-3.12.2.jar
azure-function: okhttp-3.4.2.jar
azure-function: okhttp-urlconnection-3.4.2.jar
azure-iot-edge: okhttp-3.4.2.jar
azure-iot-edge: okhttp-urlconnection-3.4.2.jar
azure-vm-agents: okhttp-3.4.2.jar
azure-vm-agents: okhttp-urlconnection-3.4.2.jar
azure-vmss: okhttp-3.14.7.jar
azure-vmss: okhttp-urlconnection-3.4.2.jar
okhttp-api: okhttp-3.12.12.jar
okhttp-api: okhttp-urlconnection-3.12.12.jar
service-fabric: okhttp-3.4.2.jar
service-fabric: okhttp-urlconnection-3.4.2.jar
upload-pgyer: okhttp-3.10.0.jar
upload-pgyer: okhttp-urlconnection-3.10.0.jar
```

### okhttp 2.x

*Low risk, but not recommended*

> We can easlily to keep okhttp 2.7.5 (including urlconnection) in the plugin for the foreseeable future. 

Reviewer: ```suggestion
We can easily to keep okhttp 2.7.5 (including urlconnection) in the plugin for the foreseeable future. 
```

These plugins could be updated to use okhttp-api-plugin with little to no risk of breakages. 
However, we recommend updating to using okhttp3 before adding this dependency.

> A few of these include okhttp-ws.  We will not be including that in the okhttp-api-plugin. We'll leave it to those projects to maintain that version on their own.  

Reviewer: Sounds good to me, but is there any particular reason for this? Is it only for okhttp 2.x, or does it pull in a bunch of additional dependencies or something?

Author: It was purely for package size and surface area, but per our other offline discussion I'm removing okhttp v2.x entirely.

This plugin manages okhttp library version for Jenkins plugins.

This plugin includes the following packages:

* `com.squareup.okio:okio` - required by `okhttp`
* `com.squareup.okhttp3:okhttp`
* `com.squareup.okhttp3:logging-interceptor` - commonly used by a number of plugins

Other `com.squareup.okhttp3.*` package may be added upon request. 

This plugin **does not** include the following packages:

* `com.squareup.okhttp:*` (all v2.x okhttp packages) - 
  The last release in this line was v2.7.5 (2016-02-25).  
  It is not longer supported in any way. 

>   Including them would only increase the risk this plugin introducing security vulnerablities and bugs.

Reviewer: ```suggestion
  Including them would only increase the risk of this plugin introducing security vulnerabilities and bugs.
```

  Plugins that want to use this plugin should update to use `okhttp3`. 
* `com.squareup.okhttp3:okhttp-urlconnection` - 

>   This package contained the `OkHttpUrlFactory` which provide a facade implementing 

Reviewer: ```suggestion
  This package contained the `OkHttpUrlFactory` class which provides a facade implementing 
```

  link:https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html[HttpURLConnection] for okhttp.
  `OkHttpUrlFactory` was deprecated in 
  link:https://square.github.io/okhttp/changelog_3x/#version-300-rc1[v3.0.0-RC1 (2016-01-02)]
  and was removed as of
  link:https://square.github.io/okhttp/changelog_3x/#version-3140[v3.14.x (2019-03-14)] 
  so there is no reason to include this package.
  Plugins that use `OkHttpUrlFactory` and want to use this plugin are strongly advised to upgrade to OkHttp's request/response API directly.

>   If that is not feasible, they can copy in a copy and paste 

Reviewer: ```suggestion
  If that is not feasible, they can copy and paste 
```

*Moderate to low risk*

link:https://square.github.io/okhttp/upgrading_to_okhttp_4/[With a few small exceptions], 
OkHttp 4.x is both binary- and Java source-compatible with OkHttp 3.x. 
You can use an OkHttp 4.x .jar file with applications or libraries built for OkHttp 3.x. 

> However, plugins that use OkHttp x4.x may depend on features not present in OkHttp 3.x.

Reviewer: ```suggestion
However, plugins that use OkHttp 4.x may depend on features not present in OkHttp 3.x.
```

This plugin does not include okhttp v2.x. 
This was done intentionally to avoid potential security issues related to bundling a version that is no longer maintained in any way.

The plugins below could be updated to use okhttp-api-plugin if they upgrade to using okhttp3.

> There is some risk of this plugin effecting the plugins below - both `okhttp` and `okhttp3` depend on `okio`.  However, `okio` is also extremely stable and is unlikely to introduce breaking changes.

Reviewer: ```suggestion
There is some risk of this plugin affecting the plugins below - both `okhttp` and `okhttp3` depend on `okio`.  However, `okio` is also extremely stable and is unlikely to introduce breaking changes.
```
